graphql_compiler/query_planning/typedefs.py

```diff
@@ -0,0 +1,295 @@
+)
+
+from .. import GraphQLDate, GraphQLDateTime, GraphQLDecimal
+from ..compiler.compiler_frontend import OutputMetadata
+from ..global_utils import is_same_type
+from ..typedefs import QueryArgumentGraphQLType
+
+
+QueryExecutionFunc = Callable[
+    [str, Dict[str, QueryArgumentGraphQLType], Dict[str, Any]], Iterable[Mapping[str, Any]]
+]
+
+# Custom scalar types.
+CUSTOM_SCALAR_TYPES = {
+    GraphQLDate.name: GraphQLDate,
+    GraphQLDateTime.name: GraphQLDateTime,
+    GraphQLDecimal.name: GraphQLDecimal,
+}
+
+# Custom scalar types must not have name conflicts with builtin scalar types.
+if set(CUSTOM_SCALAR_TYPES).intersection(specified_scalar_types):
+    raise AssertionError(
+        f"Custom scalar types must have different names than builtin scalar types. Received "
+        f"overlapping type(s) {set(CUSTOM_SCALAR_TYPES).intersection(specified_scalar_types)}. "
+        f"Custom scalar types: {set(CUSTOM_SCALAR_TYPES)}. Builtin scalar types: "
+        f"{set(specified_scalar_types)}."
+    )
+
+# Custom scalar types combined with builtin scalar types represent all allowable scalar types.
+ALL_SCALAR_TYPES = copy.copy(CUSTOM_SCALAR_TYPES)
+ALL_SCALAR_TYPES.update(specified_scalar_types)
+
+
+def _type_from_scalar_type_dictionary(
+    scalar_types: Dict[str, GraphQLScalarType], type_node: TypeNode
+) -> GraphQLType:
+    """Get the GraphQL type definition from an AST node.
+
+    Given a scalar type dictionary and an AST node describing a type, return a GraphQLType
+    definition, which applies to that type. For example, if provided the parsed AST node for
+    `[Date]`, a GraphQLList instance will be returned, containing the type called
+    "Date" found in the scalar type dictionary. If a type called "Date" is not found in the scalar
+    type dictionary, then None will be returned.
+
+    Note: this is very similar to GraphQL's type_from_ast. However, instead of requiring a GraphQL
+    schema this function requires a dictionary of the scalar types. This simplifies deserialization
+    and allows for custom scalar types without constructing an entire schema.
+
+    Args:
+        scalar_types: dictionary mapping type name to GraphQLScalarType
+        type_node: AST node describing a type
+
+    Returns:
+        GraphQLType that applies to the type specified in type_node.
+
+    Raises:
+        AssertionError: if an invalid type node is given.
+    """
+    if isinstance(type_node, ListTypeNode):
+        inner_type = _type_from_scalar_type_dictionary(scalar_types, type_node.type)
+        if inner_type:
+            return GraphQLList(inner_type)
+        else:
+            raise AssertionError(
+                f"Invalid type node. ListTypeNode contained inner type {inner_type}."
+            )
+    elif isinstance(type_node, NonNullTypeNode):
+        inner_type = _type_from_scalar_type_dictionary(scalar_types, type_node.type)
+        if inner_type:
+            return GraphQLNonNull(inner_type)
```

Comment:
There's a mypy error here that I'm not sure how to fix.

`graphql_compiler/query_planning/typedefs.py:90: error: Value of type variable "GNT" of "GraphQLNonNull" cannot be "GraphQLType"`

Here's the relevant `GraphQLNonNull` [code](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/type/definition.py#L1505). It is expecting a [GraphQLNullableType](https://github.com/graphql-python/graphql-core/blob/main/src/graphql/type/definition.py#L1559), which is a Union of other GraphQL types. I know we have [is_same_type function](https://github.com/kensho-technologies/graphql-compiler/blob/c137fa06c7f583889202068302405e5edb3e681b/graphql_compiler/global_utils.py#L88) for comparing GraphQL types, but I'm not sure how type comparison should work when there is a `Union` of multiple GraphQL types.